gh-130149: refactor tests for HMAC

picnixz opened the pull request:
- [x] test the `repr()` of HMAC OpenSSL objects properly.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-130149 -->
* Issue: gh-130149
<!-- /gh-issue-number -->

picnixz committed:
add helper for _hashlib detection

picnixz committed:
remove unused function `ignore_warning`

picnixz committed:
cleanup whitespaces

picnixz committed:
clean imports and test guards

- Use `_hashlib` instead of `_hashopenssl` naming.
- Use `_hashlib.*` instead of top-level imports.
- Consistently use new hashlib helper guards.

picnixz committed:
Add mixins for interacting with HMAC

picnixz committed:
add mixin for RFC test cases

picnixz committed:
use RFC mixin classes

picnixz committed:
refactor `CompareDigestTestCase`

picnixz committed:
refactor `CopyTestCase`

picnixz committed:
refactor `UpdateTestCase`

picnixz committed:
refactor `SanityTestCase`

picnixz committed:
refactor `ConstructorTestCase`

picnixz committed:
Merge branch 'main' into test/hmac/refactor-99108

picnixz committed:
update comments

picnixz committed:
extend coverage

picnixz committed:
Merge branch 'main' into test/hmac/refactor-99108

picnixz committed:
add regression test for HMAC `repr()`

picnixz committed:
change a bit the digestmod tests

picnixz committed:
Merge branch 'main' into test/hmac/refactor-99108

gpshead reviewed (approved):
While I'm not a fan of a giant maze of mixin classes from a code clarify point of view (nobody reading code can rationally follow a non-trivial class hierarchy MRO graph in their head)... this refactor should actually work and allow for easier testing of our too many possible implementation choices - and I don't have meaningfully better ideas for that that'd work within the stdlib test suite.

picnixz committed:
remove mixin that is not yet needed

picnixz committed:
add comments

picnixz committed:
semantics: 'digest' -> 'hexdigest'

picnixz committed:
use `__init_subclass__` instead of `setUpClass`

picnixz committed:
document and reorganize method order in `DigestModTestCaseMixin`

picnixz committed:
use separate factories for creating bad digestmod cases

One factory is for test cases when the digestmod parameter is missing;
the other is for test cases when the digestmod parameter is unknown.

picnixz committed:
minor tweak on `CompareDigestMixin`

picnixz committed:
do not allow 'hashfunc' or 'hashname' to be None for RFC tests

I previously allowed it to be None for future compatibility, but
I think it's better to construct a fake hash function instead, or
